### ISSUE TITLE
Error when trying to setup file synchronization with Google Drive

### ISSUE THREAD

**Issue by ulfskogeng** [milan-fabian/slideshow-support#34]
I'm getting "this page isn't working host: 127.0.0.1 Error Code: net::ERR_CONNECTION_REFUSED" when I try to setup Google Drive File synchronization. Is this a know error? Using Puffin Web browser
![IMG_20230120_144950](https://user-images.githubusercontent.com/18708500/213715403-ac20f220-69cc-4829-b736-9924c0b59b6a.jpg)

**Comment by milan-fabian** [comment]
@ulfskogeng Can you please specify in which step of the Google Drive File synchronization does this error happen?

**Comment by ulfskogeng** [comment]
@milan-fabian It's right after I click "Continue" on this screen 
![IMG_20230120_154558](https://user-images.githubusercontent.com/18708500/213727095-05bcf872-fe39-4687-8d5f-a72c2f024608.jpg)

**Comment by milan-fabian** [comment]
@ulfskogeng Please make sure Slideshow's web interface is running either on port 8080 or 80, other ports are unfortunately not supported for login to Google Drive.
If the web interface is running on one of these two ports, please reboot the entire device and try again. If that won't help either, please send a bug report via menu Information - Bug report, it will automatically attach logs, I will check them.

**Comment by ulfskogeng** [comment]
web interface is running on port 8080. Bug report submitted after reboot. Thank you @milan-fabian !

**Comment by milan-fabian** [comment]
@ulfskogeng Can you please make sure the entire device is rebooted before trying again? I can see in the logs that the uptime is 6 hours. The best way would be to unplug the Chromecast from power for 10 seconds and plug it in again.
There might be something incorrectly initialized, that's why it's best to reboot it.

**Comment by ulfskogeng** [comment]
@milan-fabian I think the problem is due to the fact that Chrome Web browser is not installed on Google TV. Sideloading is possible of course, but trying to navigate is not possible, and therefore you cannot complete the steps to verify the connection to GDrive. 

I have tried various other browsers, but they cannot be used as they are not compliant and that's probably the reason for the error message as reported by me. 

Everything is working fine on an Android phone, so I tried to make a backup of the slideshow folder and restore it on the Chromecast, but the file sync. does not work even if all the settings are restored. Is it possible to make a fix for this?

**Comment by milan-fabian** [comment]
@ulfskogeng Google Chrome browser is preferred when logging in to Google Drive account, but unless you are getting "Error 403: disallowed_useragent" page, using a different browser should be fine.

Based on your description, the error happens at the moment Google Drive is redirecting back to Slideshow app with the access token, but apparently Slideshow app is inaccessible ("ERR_CONNECTION_REFUSED"). This can be cause by either:

- It takes too long to log in into Google Drive account (more than 3 minutes) - according to the logs, this doesn't seem to be an issue in your case.
- There is some kind of mismatch with HTTP port - in best case, Slideshow web interface should be running on port 8080 and Google Drive should redirect to port 8080 after logging in (you can verify it by checking the URL address when "ERR_CONNECTION_REFUSED" message is displayed).

Google Drive access token is stored securely, usually in folder "/data/data/sk.mimac.slideshow/app_data/", where it is not accessible to other apps, so it's not transferred when you copy "slideshow" folder, nor it is included in the configuration backup. This is due to Google Drive security policy, the token might be revoked if it is used from multiple devices.

**Comment by ulfskogeng** [comment]
@milan-fabian
Uninstalled your App
Deleted the slideshow folder
Installed your App again. 
It starts up on port 8080 even if the port in settings.xml is set to port 80
I have rebooted and unplugged the Chromecast/ Android TV to make sure it's not cached or something like that. For some reason it does not read the port settings from the config file
![IMG_20230122_123733](https://user-images.githubusercontent.com/18708500/213913946-e168ff0e-13a2-4775-908a-dfcaed509e41.jpg)
![IMG_20230122_123801](https://user-images.githubusercontent.com/18708500/213913950-90307897-1bba-4da6-b853-5abcb818389d.jpg)

**Comment by milan-fabian** [comment]
@ulfskogeng On Chromecast the web interface will be accessible only on port 8080 (or a port higher than 1000, if you set it), this is expected. When the web interface is running on port 8080, there should be no problem with logging to Google Drive.

The reason why the web interface can't run on port 80 on Chromecast is that the device is not rooted and regular applications on non-rooted Android can't bind to a port lower than 1000, due to security restrictions of the Linux-based operating system. You can find more details about this on https://slideshow.digital/2020/05/network-port-numbers-on-android/.

**Comment by ulfskogeng** [comment]
I finally managed to make it work. I found a browser that did not give me the error message
![IMG_20230124_084959](https://user-images.githubusercontent.com/18708500/214402999-fab5bc80-cde6-4267-8089-5b4db24a238a.jpg)
Everything is working beautifully. Thank you so much for making this excellent app @milan-fabian

**Comment by milan-fabian** [comment]
I am glad it works now :)

**Comment by ulfskogeng** [comment]
@milan-fabian Google Drive synchronization stopped working after upgrading to ver 4.2.2 I have just sent you a bug report. If I try to setup Google Drive File synchronization it will not display the normal Google authentication webpage.

**Comment by milan-fabian** [comment]
@ulfskogeng Unfortunately, there was a bug in version 4.2.0 which caused saving the Google Drive token in a bad format. This was fixed in Slideshow 4.2.2, but in some cases, manual fix might be needed as well. This can be done by opening Slideshow's web interface - menu Settings - Shell console, entering the command below and clicking on Send.
```
rm /data/data/sk.mimac.slideshow/app_data/tokens/StoredCredential
```
After executing this command, you should be able to re-login to your Google Drive account via File synchronization page.

**Comment by ulfskogeng** [comment]
@milan-fabian your solution worked as advertised. Thx for excellent support!
